File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        lowest = prices[0]
        profit = 0

        for price in prices:
            lowest = min(lowest, price)
            profit = max(profit, price - lowest)
        
        return profit
        

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int buyingPrice = prices[0];

        for (int sellingPrice: prices){
            buyingPrice = Math.min(buyingPrice, sellingPrice);
            maxProfit = Math.max(maxProfit, sellingPrice - buyingPrice);
        }

        return maxProfit;
    }
}